File: java/believe/tools/ProtoFileSerializer.java
package believe.tools;

import com.google.protobuf.Message;
import com.google.protobuf.TextFormat;
import com.google.protobuf.TextFormat.ParseException;
import org.newdawn.slick.util.Log;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.InvocationTargetException;
import java.util.Scanner;

/**
 * Takes a proto message reference and an arbitrary number of textproto files and serializes the
 * textprotos into corresponding {@code *.txt} files.
 *
 *
 * <p>Usage:
 * <pre>
 *   java -jar proto_file_serializer.jar com.proto.MessageType proto1.textproto proto2.textproto
 * </pre>
 */
public class ProtoFileSerializer {
  private static final String
      USAGE =
      "Usage:\n"
          + "    bazel-out/.../proto_file_serializer com.proto.MessageType /path/to/out/dir "
          + "proto1.textproto [other text protos]";

  public static void main(String[] args) throws
      IOException,
      ClassNotFoundException,
      NoSuchMethodException,
      InvocationTargetException,
      IllegalAccessException {
    if (!verifyArgs(args)) {
      return;
    }

    Class<?> protoClass = Class.forName(args[0]);
    if (!(Message.class.isAssignableFrom(protoClass))) {
      throw new IllegalArgumentException("Class '" + args[0] + "' is not a proto message.");
    }

    String outputDir = args[1];
    ProtoFileSerializer serializer = new ProtoFileSerializer();

    for (int i = 2; i < args.length; i++) {
      String arg = args[i];
      Message.Builder builder = (Message.Builder) protoClass.getMethod("newBuilder").invoke(null);

      File inputFile = new File(arg);
      if (!inputFile.exists()) {
        Log.error("Could not find file '" + arg + "'. Skipping...");
        continue;
      }

      String fileName = arg.substring(arg.lastIndexOf("/") + 1, arg.lastIndexOf(".")) + ".pb";
      String filePath = outputDir + "/" + fileName;
      File outputFile = new File(filePath);
      try (
          FileInputStream textProtoInputStream = new FileInputStream(inputFile);
          FileOutputStream binaryProtoOutputStream = new FileOutputStream(outputFile)) {
        serializer.convert(builder, textProtoInputStream, binaryProtoOutputStream);
      } catch (ParseException e) {
        Log.error("Could not read '" + arg + "' as " + args[0] + ". Skipping...", e);
      }
    }
  }

  private static boolean verifyArgs(String[] args) {
    if (args.length < 3) {
      System.out.println(USAGE);
      return false;
    }
    return true;
  }

  void convert(
      Message.Builder builder,
      InputStream textProtoInputStream,
      OutputStream binaryProtoOutputStream) throws IOException {
    Scanner s = new Scanner(textProtoInputStream).useDelimiter("\\A");
    TextFormat.getParser().merge(cb -> {
      if (!s.hasNext()) {
        return -1;
      }
      String result = s.next();
      cb.append(result);
      return result.length();
    }, builder);
    builder.build().writeTo(binaryProtoOutputStream);
  }
}

File: java/believe/tools/ProtoFileSerializer.kt
package believe.tools

import believe.proto.TextProtoParser
import com.google.protobuf.Message
import dagger.Reusable
import org.newdawn.slick.util.Log
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import javax.inject.Inject

/**
 * Takes a proto message reference and an arbitrary number of textproto files and serializes the
 * textprotos into corresponding `*.txt` files.
 *
 * Usage:
 * ```
 * java -jar proto_file_serializer.jar \
 *     com.proto.MessageType \
 *     proto1.textproto \
 *     proto2.textproto
 * ```
 */
@Reusable
class ProtoFileSerializer @Inject constructor(
    private val textProtoParserFactory: TextProtoParser.Factory
) {

    @Throws(ClassNotFoundException::class, IOException::class)
    fun serialize(args: Array<String>) {
        val protoClass = Class.forName(args[0])
        require(Message::class.java.isAssignableFrom(protoClass)) {
            "Class '" + args[0] + "' is not a proto message."
        }
        // Safe cast
        @Suppress("UNCHECKED_CAST") val textProtoParser =
            textProtoParserFactory.create(protoClass as Class<out Message>)
        val outputDir = args[1]
        for (i in 2 until args.size) {
            val arg = args[i]
            val inputFile = File(arg)
            if (!inputFile.exists()) {
                Log.error("Could not find file '$arg'. Skipping...")
                continue
            }
            val fileName = arg.substring(arg.lastIndexOf("/") + 1, arg.lastIndexOf(".")) + ".pb"
            val filePath = "$outputDir/$fileName"
            val outputFile = File(filePath)
            FileInputStream(inputFile).use { textProtoInputStream ->
                FileOutputStream(outputFile).use { binaryProtoOutputStream ->
                    val parsedProto = textProtoParser.parse(textProtoInputStream)
                    if (parsedProto == null) {
                        Log.error("Failed to generate $fileName due to a parsing error.")
                    } else {
                        parsedProto.writeTo(binaryProtoOutputStream)
                    }
                }
            }
        }
    }
}